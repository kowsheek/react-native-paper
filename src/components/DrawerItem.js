/* @flow */

import color from 'color';
import * as React from 'react';
import { View, Text, StyleSheet } from 'react-native';
import Icon from './Icon';
import TouchableRipple from './TouchableRipple';
import withTheme from '../core/withTheme';
import type { Theme } from '../types';
import type { IconSource } from './Icon';

type Props = {
  /**
   * The label text of the item.
   */
  label: string,
  /**
   * Icon to display for the `DrawerItem`.
   */
  icon?: IconSource,
  /**
   * Whether to highlight the drawer item as active.
   */
  active?: boolean,
  /**
   * Function to execute on press.
   */
  onPress?: () => mixed,
  /**
   * @optional
   */
  theme: Theme,
};

/**
 * DrawerItem is a component used to show an action item with an icon and a label in a navigation drawer.
 *
 * ## Usage
 * ```js
 * import * as React from 'react';
 * import { DrawerItem } from 'react-native-paper';
 *
 * const MyComponent = () => (
 *   <DrawerItem label="First Item" />
 * );
 * ```
 */
class DrawerItem extends React.Component<Props> {
  render() {
    const { icon, label, active, theme, ...props } = this.props;
    const { colors, roundness } = theme;
    const backgroundColor = active
      ? color(colors.primary)
          .alpha(0.12)
          .rgb()
          .string()
      : 'transparent';
    const contentColor = active
      ? colors.primary
      : color(colors.text)
          .alpha(0.68)
          .rgb()
          .string();
    const fontFamily = theme.fonts.medium;
    const labelMargin = icon ? 32 : 0;

    return (
      <TouchableRipple
        {...props}
        style={[styles.touchable, { borderRadius: roundness }]}
      >
        <View
          style={[styles.wrapper, { backgroundColor, borderRadius: roundness }]}
        >
          {icon && <Icon source={icon} size={24} color={contentColor} />}
          <Text
            numberOfLines={1}
            style={{
              color: contentColor,
              fontFamily,
              marginLeft: labelMargin,
              marginRight: 32,
            }}
          >
            {label}
          </Text>
        </View>
      </TouchableRipple>
    );
  }
}

const styles = StyleSheet.create({
  touchable: {
    marginHorizontal: 10,
    marginVertical: 4,
  },
  wrapper: {
    flexDirection: 'row',
    alignItems: 'center',
    padding: 8,
    height: 40,
  },
});

export default withTheme(DrawerItem);
