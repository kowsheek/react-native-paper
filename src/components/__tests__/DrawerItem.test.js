/* @flow */

import * as React from 'react';
import renderer from 'react-test-renderer';
import DrawerItem from '../DrawerItem';

it('renders basic DrawerItem', () => {
  const tree = renderer
    .create(<DrawerItem onPress={() => {}} label="Example item" />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('renders DrawerItem with icon', () => {
  const tree = renderer
    .create(<DrawerItem icon="info" label="Example item" />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('renders active DrawerItem', () => {
  const tree = renderer
    .create(<DrawerItem icon="info" active label="Example item" />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});
