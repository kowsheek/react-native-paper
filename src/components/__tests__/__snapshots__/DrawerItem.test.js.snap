// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders DrawerItem with icon 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  hasTVPreferredFocus={undefined}
  hitSlop={undefined}
  isTVSelectable={true}
  nativeID={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "marginHorizontal": 10,
        "marginVertical": 4,
      },
      Object {
        "borderRadius": 4,
      },
    ]
  }
  testID={undefined}
  tvParallaxProperties={undefined}
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "height": 40,
          "padding": 8,
        },
        Object {
          "backgroundColor": "transparent",
          "borderRadius": 4,
        },
      ]
    }
  >
    <Text
      accessible={true}
      allowFontScaling={false}
      ellipsizeMode="tail"
      pointerEvents="none"
      style={
        Array [
          Object {
            "color": "rgba(0, 0, 0, 0.68)",
            "fontSize": 24,
          },
          Object {
            "backgroundColor": "transparent",
          },
          Object {
            "fontFamily": "Material Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
        ]
      }
    >
      
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      numberOfLines={1}
      style={
        Object {
          "color": "rgba(0, 0, 0, 0.68)",
          "fontFamily": "HelveticaNeue-Medium",
          "marginLeft": 32,
          "marginRight": 32,
        }
      }
    >
      Example item
    </Text>
  </View>
</View>
`;

exports[`renders active DrawerItem 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  hasTVPreferredFocus={undefined}
  hitSlop={undefined}
  isTVSelectable={true}
  nativeID={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "marginHorizontal": 10,
        "marginVertical": 4,
      },
      Object {
        "borderRadius": 4,
      },
    ]
  }
  testID={undefined}
  tvParallaxProperties={undefined}
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "height": 40,
          "padding": 8,
        },
        Object {
          "backgroundColor": "rgba(98, 0, 238, 0.12)",
          "borderRadius": 4,
        },
      ]
    }
  >
    <Text
      accessible={true}
      allowFontScaling={false}
      ellipsizeMode="tail"
      pointerEvents="none"
      style={
        Array [
          Object {
            "color": "#6200ee",
            "fontSize": 24,
          },
          Object {
            "backgroundColor": "transparent",
          },
          Object {
            "fontFamily": "Material Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
        ]
      }
    >
      
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      numberOfLines={1}
      style={
        Object {
          "color": "#6200ee",
          "fontFamily": "HelveticaNeue-Medium",
          "marginLeft": 32,
          "marginRight": 32,
        }
      }
    >
      Example item
    </Text>
  </View>
</View>
`;

exports[`renders basic DrawerItem 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  hasTVPreferredFocus={undefined}
  hitSlop={undefined}
  isTVSelectable={true}
  nativeID={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "marginHorizontal": 10,
        "marginVertical": 4,
      },
      Object {
        "borderRadius": 4,
      },
    ]
  }
  testID={undefined}
  tvParallaxProperties={undefined}
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "height": 40,
          "padding": 8,
        },
        Object {
          "backgroundColor": "transparent",
          "borderRadius": 4,
        },
      ]
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      numberOfLines={1}
      style={
        Object {
          "color": "rgba(0, 0, 0, 0.68)",
          "fontFamily": "HelveticaNeue-Medium",
          "marginLeft": 0,
          "marginRight": 32,
        }
      }
    >
      Example item
    </Text>
  </View>
</View>
`;
